File: Excel/484 Challenge.R
library(tidyverse)
library(readxl)
library(gmp)

path = "Excel/484 Pythagorean Triplets for a Sum.xlsx"
input = read_xlsx(path, range = "A2:A10")
test  = read_xlsx(path, range = "B2:D10") %>% 
  mutate(across(everything(), as.numeric))

find_pythagorean_triplet <- function(P) {
  m_max <- floor(sqrt(P / 2))
  
  possible_values <- expand_grid(m = 2:m_max, n = 1:(m_max - 1)) %>%
    filter(m > n, m %% 2 != n %% 2, gcd(m, n) == 1)
  
  triplets <- possible_values %>%
    pmap(function(m, n) {
      k <- P / (2 * m * (m + n))
      if (k == floor(k)) {
        a <- k * (m^2 - n^2)
        b <- k * 2 * m * n
        c <- k * (m^2 + n^2)
        return(c(a, b, c))
      } else {
        return(NULL)
      }
    })
  
  triplet <- triplets %>%
    compact() %>%
    keep(~ sum(.x) == P)
  
  if (length(triplet) > 0) {
    result <- triplet[[1]]
  } else {
    result <- c(NA_real_, NA_real_, NA_real_)
  }
  
  tibble(a = result[1], b = result[2], c = result[3])
}

result = input %>%
  pmap_dfr(~ find_pythagorean_triplet(..1))

# in one case (for 132) I get another but correct result

File: Excel/400-499/484/484 Challenge.py
import math
import itertools
import pandas as pd

path = "484 Pythagorean Triplets for a Sum.xlsx"
input = pd.read_excel(path, usecols="A", skiprows=1)
test = pd.read_excel(path, usecols="B:D", skiprows=1)

def find_pythagorean_triplet(P):
    m_max = math.floor(math.sqrt(P / 2))
    
    possible_values = list(itertools.product(range(2, m_max + 1), range(1, m_max)))
    possible_values = [values for values in possible_values if values[0] > values[1] and values[0] % 2 != values[1] % 2 and math.gcd(values[0], values[1]) == 1]
    
    triplets = []
    for m, n in possible_values:
        k = P / (2 * m * (m + n))
        if k == math.floor(k):
            a = k * (m**2 - n**2)
            b = k * 2 * m * n
            c = k * (m**2 + n**2)
            triplets.append([a, b, c])
    
    triplet = [triplet for triplet in triplets if sum(triplet) == P]
    
    if len(triplet) > 0:
        result = triplet[0]
    else:
        result = [float('nan'), float('nan'), float('nan')]
    
    return pd.Series(result, index=['a', 'b', 'c'])

output = input['Sum'].apply(find_pythagorean_triplet)

# one case (for 132), I get different but correct answer